nilmtk/api.py

```diff
@@ -208,36 +209,93 @@ def train_jointly(self,clf,d):
-                    appliance_df = next(train.buildings[building].elec[appliance_name].load(physical_quantity='power', ac_type=self.power['appliance'], sample_period=self.sample_period))
-                    appliance_df = appliance_df[[list(appliance_df.columns)[0]]]
-                    appliance_readings.append(appliance_df)
 
-                if self.DROP_ALL_NANS:
-                    train_df, appliance_readings = self.dropna(train_df, appliance_readings)
+                if self.use_activations:
+                    print("Getting appliance activations")
+                    for appliance_name in self.appliances:
+                        train.set_window(start=d[dataset]['buildings'][building]['start_time'],
+                                         end=d[dataset]['buildings'][building]['end_time'])
+
+                        activations = train.buildings[building].elec[appliance_name].activation_series(
+                                self.appliances_params[appliance_name]["min_off_time"], 
+                                self.appliances_params[appliance_name]["min_on_time"], 
+                                self.appliances_params[appliance_name]["number_of_activation_padding"], 
+                                self.appliances_params[appliance_name]["min_on_power"])
+                            
+                        for period in activations:
+                            timestamp = period.index[0]
+                            tz_offset = (timestamp.replace(tzinfo=None) - 
+                                        timestamp.tz_convert('UTC').replace(tzinfo=None))
+                            tz_offset_bg = np.timedelta64(tz_offset)
+
+                            timestamp = period.index[-1]
+                            tz_offset = (timestamp.replace(tzinfo=None) - 
+                                        timestamp.tz_convert('UTC').replace(tzinfo=None))
+
+                            tz_offset_end = np.timedelta64(tz_offset)
+
+                            if tz_offset_bg > tz_offset_end:
+                                beg = period.index[0].tz_localize(None) - tz_offset_bg
+                                end = period.index[-1].tz_localize(None)
+                            elif tz_offset_bg < tz_offset_end: 
+                                beg = period.index[0].tz_localize(None)
+                                end = period.index[-1].tz_localize(None) - tz_offset_end 
+                            else:
+                                beg = period.index[0].tz_localize(None)
+                                end = period.index[-1].tz_localize(None)
+
+                            train.set_window(
+                                    start=beg,
+                                    end=end
+                                    )
+
+                            mains_df = next(train.buildings[building].elec.mains().load(physical_quantity='power', ac_type=self.power['mains'], sample_period=self.sample_period))
```

Comment:
Instead of the very complicated logic above involving `set_window`, why don't you load the mains data before the for loop and use index-based selection?
Something like:
```python3
mains_df = next(train.buildings[building].elec.mains().load(...))
for event in activations: # note that period is a very confusing name, event would be better.
    mains_event = mains_df.loc[event.index]
    if not mains_event.empty:
        # the code goes on, using mains_event instead of mains_df.
```